server/backends/chunkstore/chunkstore.go

```diff
@@ -0,0 +1,413 @@
+		blobName:   blobName,
+		ctx:        ctx,
+		reverse:    true,
+	}, nil
+}
+
+type writeResult struct {
+	err  error
+	size int
+}
+
+type chunkstoreWriter struct {
+	ctx                context.Context
+	chunkstore         *Chunkstore
+	writeChannel       chan []byte
+	writeResultChannel chan writeResult
+	blobName           string
+	blockSize          int
+	timeoutDuration    time.Duration
+}
+
+func (w *chunkstoreWriter) Write(p []byte) (int, error) {
+	w.writeChannel <- p
+	select {
+	case result, open := <-w.writeResultChannel:
+		if !open {
+			return 0, os.ErrClosed
+		}
+		return result.size, result.err
+	case <-time.After(20 * time.Second):
+		return 0, context.DeadlineExceeded
+	}
+}
+
+func (w *chunkstoreWriter) Flush() (int, error) {
+	w.writeChannel <- nil
+	select {
+	case result, open := <-w.writeResultChannel:
+		if !open {
+			return 0, os.ErrClosed
+		}
+		return result.size, result.err
+	case <-time.After(20 * time.Second):
+		return 0, context.DeadlineExceeded
+	}
+}
+
+func (w *chunkstoreWriter) Close() error {
+	close(w.writeChannel)
+	select {
+	case result, open := <-w.writeResultChannel:
+		if !open {
+			return fmt.Errorf("Error closing %v: %w", w.blobName, os.ErrClosed)
+		}
+		return result.err
+	case <-time.After(w.timeoutDuration):
+		return os.ErrDeadlineExceeded
+	}
+}
+
+func (w *chunkstoreWriter) writeLoop() {
+	chunkIndex := uint16(0)
+	chunk := []byte{}
+	never := time.Unix(1<<63-62135596801, 999999999)
+	flushTime := never
+	var writeError error
+	bytesFlushed := 0
+	closed := false
+	for !closed {
+		timeout := false
+		select {
+		case p, open := <-w.writeChannel:
+			if !open {
+				closed = true
+			}
+			if open && p == nil {
+				flushTime = time.Unix(0, 0)
+			} else {
+				chunk = append(chunk, p...)
+			}
+		case <-time.After(time.Until(flushTime)):
+			timeout = true
+		case <-w.ctx.Done():
+			closed = true
+		}
+		for len(chunk) >= w.blockSize {
+			bytesWritten, err := w.chunkstore.writeChunk(w.ctx, w.blobName, chunkIndex, chunk[:w.blockSize])
+			bytesFlushed += bytesWritten
+			chunk = chunk[bytesWritten:]
+			if bytesWritten > 0 {
+				chunkIndex++
+			}
+			if err != nil {
+				writeError = err
+				break
+			}
+		}
+		if writeError == nil && ((closed && chunkIndex == 0) || (len(chunk) > 0 && (time.Now().After(flushTime) || closed))) {
+			fmt.Printf("Flush time is %v, Flushing %v\n", flushTime, chunk)
+			bytesWritten, err := w.chunkstore.writeChunk(w.ctx, w.blobName, chunkIndex, chunk)
+			bytesFlushed += bytesWritten
+			chunk = chunk[bytesWritten:]
+			if bytesWritten > 0 {
+				chunkIndex++
+			}
+			if err != nil {
+				writeError = err
+			}
+		}
+		if len(chunk) == 0 {
+			flushTime = never
+		} else if time.Until(flushTime) > (w.timeoutDuration) {
+			flushTime = time.Now().Add(w.timeoutDuration)
+		}
+		if timeout {
+			continue
+		}
+		w.writeResultChannel <- writeResult{size: bytesFlushed, err: writeError}
+		writeError = nil
+		bytesFlushed = 0
+	}
+	close(w.writeResultChannel)
```

Comment:
There's a lot of logic here -- could see move some of these blocks into functions and then have this writeLoop just be the control flow?